### PR TITLE
Converting graph fmt.Errorf to new error code package format

### PR DESCRIPTION
**Pull request by mikebrow**
Convert "graph" static error strings to the new error code package format

Signed-off-by: Mike Brown brownwm@us.ibm.com

### CODE REVIEW

File: errors/graph.go [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+	"github.com/docker/distribution/registry/api/errcode"
+)
+
+var (
+	// ErrorCodeNoSuchImage is generated when we can't find the image.
+	//
+	ErrorCodeNoSuchImage = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "NOSUCHIMAGE",
+		Message:        "could not find image: %v",
```
Comment (duglin) [inline errors/graph.go]: Please add a "Description" field to each.

File: errors/graph.go [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+import (
+	"net/http"
+
+	"github.com/docker/distribution/registry/api/errcode"
+)
+
+var (
+	// ErrorCodeNoSuchImage is generated when we can't find the image.
+	//
```
Comment (duglin) [inline errors/graph.go]: remove

File: errors/graph.go [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+	ErrorCodeNoSuchImage = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "NOSUCHIMAGE",
+		Message:        "could not find image: %v",
+		HTTPStatusCode: http.StatusNotFound,
+	})
+
+	// ErrorCodeImageIDConflict is generated when the expected id does not
+	// match the id of the loaded image.
+	ErrorCodeImageIDConflict = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "IMAGEWITHWRONGIMAGEID",
```
Comment (duglin) [inline errors/graph.go]: Pleas make sure the Value field is the same as the ErrorCode variable, just without the "ErrorCode" part and the rest of it uppercase. e.g.  `IMAGEIDCONFLICT` in this case.

File: errors/graph.go (hunk 1 above)
Comment (mikebrow) [inline errors/graph.go]: Done..

File: errors/graph.go (hunk 2 above)
Comment (mikebrow) [inline errors/graph.go]: Done...

File: errors/graph.go (hunk 3 above)
Comment (mikebrow) [inline errors/graph.go]: Done...

**Comment by duglin** [comment]
LGTM

File: errors/graph.go [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+		Description:    "Failure writing image size to graph image directory",
+		HTTPStatusCode: http.StatusInternalServerError,
+	})
+
+	// ErrorCodeBadDigestWriteOP is generated when the attempt to write the
+	// digest for the image has failed.
+	ErrorCodeBadDigestWriteOP = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "BADDIGESTWRITEOP",
+		Message:        "Error storing digest in %s/%s: %s",
+		Description:    "Failture writing digest for an image",
```
Comment (abronan) [inline errors/graph.go]: Failture -> Failure

File: errors/graph.go [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+		Description:    "The key does not have the required read/write permission",
+		HTTPStatusCode: http.StatusForbidden,
+	})
+
+	// ErrorCodeWriteImageToFSFailed is generated when a failure occurs verifying
+	// the writting of the image to the FS.
+	ErrorCodeWriteImageToFSFailed = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "WRITEIMAGETOFSFAILED",
+		Message:        "image verification failed for digest %s",
+		Description:    "The writting of the image to the file system and subsequent digest verification has failed",
```
Comment (abronan) [inline errors/graph.go]: writting -> writing

File: errors/graph.go [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+		Description:    "The generation of a JSON object has failed for the specified image at the specified path",
+		HTTPStatusCode: http.StatusConflict,
+	})
+
+	// ErrorCodeGenerateLayer is generated when the generation of the layer
+	// archive fails.
+	ErrorCodeGenerateLayer = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "GENERATELAYER",
+		Message:        "Failed to generate layer archive: %s",
+		Description:    "The specified error resulted from attemtping to generate a layer archive",
```
Comment (abronan) [inline errors/graph.go]: attemtping -> attempting

**Comment by abronan** [comment]
Spotted minor typos in some `Description` fields, but LGTM when addressed :)

**Comment by mikebrow** [comment]
Thx @abronan, I've corrected the typos.

**Comment by abronan** [comment]
Thanks @mikebrow!

ping @dmcgowan @RichardScothern

**Comment by abronan** [comment]
@stevvooe I think you had concerns on another PR related to the formatting and the usage of the `errcode` package. Does this apply here as well?

**Comment by stevvooe** [comment]
@abronan This PR has the exact same problem. I really cannot see how adding 50+ error codes helps usability.

**Comment by LK4D4** [comment]
@stevvooe I agree.

**Comment by duglin** [comment]
@stevvooe please see my email and we can discuss.  However, Docker today generates a certain number of unique error messages, if we keep **the same** UX as today, I don't see how we don't then have to have the same number of error codes (or hand-crafted mapping from one errorCode to many error messages which a lot more work and brittle).

**Comment by abronan** [comment]
@stevvooe After looking back at the PR I tend to agree..

Putting this back to design-review to make sure we agree on the plan here.

But if the usage of the pkg is far off from what it was designed for, then we should document the pkg properly (docker/distribution#1030) and close this one.

**Comment by LK4D4** [comment]
@mikebrow can we close this?

**Comment by mikebrow** [comment]
Yes please.. I'll fix and resubmit after we have the final plan/design.

**Pull request closed without merging**